pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,175 @@
+    """Calls diagonalize_observables on each observable in a list. After each observable is diagonalized,
+    if diagonalizing gates were added, its wire is removed from the list of wires still needing
+    diagonalization. This prevents diagonalizing twice on the same wire for overlapping measurements, like
+    [qml.X(0), qml.X(0)@qml.Y(1)]"""
+
+    new_obs = []
+    diagonalizing_gates = []
+
+    for o in obs_list:
+        gates, obs = diagonalize_observable(o, diagonalization_wires, unsupported_obs)
+        if gates:
+            diagonalizing_gates.extend(gates)
+            for w in obs.wires:
+                diagonalization_wires.remove(w)
+        new_obs.append(obs)
+
+    return diagonalizing_gates, new_obs
+
+
+@singledispatch
+def _diagonalize_complex_observable(
+    observable, diagonalization_wires, unsupported_obs=_default_unsupported_obs
+):
+    """takes an observable consisting of multiple other observables, and changes all
+    unsupported obs to the measurement basis. Applies diagonalizing gates unless the
+    observable being diagonalized isn't on one of the diagonalization wires."""
+
+    raise NotImplementedError(
+        f"Unable to convert observable of type {type(observable)} to the measurement basis"
+    )
+
+
+@_diagonalize_complex_observable.register
+def _diagonalize_symbolic_op(
+    observable: SymbolicOp, diagonalization_wires, unsupported_obs=_default_unsupported_obs
+):
+    diagonalizing_gates, new_base = diagonalize_observable(
+        observable.base, diagonalization_wires, unsupported_obs
+    )
+
+    new_observable = copy(observable)
+    new_observable._hyperparameters["base"] = new_base
+
+    return diagonalizing_gates, new_observable
+
+
+@_diagonalize_complex_observable.register
+def _diagonalize_tensor(
+    observable: Tensor, diagonalization_wires, unsupported_obs=_default_unsupported_obs
+):
+    diagonalizing_gates, new_obs = get_obs_and_gates(
+        observable.obs, diagonalization_wires, unsupported_obs
+    )
+
+    new_observable = copy(observable)
+    new_observable.obs = new_obs
+
+    return diagonalizing_gates, new_observable
+
+
+@_diagonalize_complex_observable.register
+def _diagonalize_hamiltonian(
+    observable: qml.ops.Hamiltonian,
+    diagonalization_wires,
+    unsupported_obs=_default_unsupported_obs,
+):
+    diagonalizing_gates, new_ops = get_obs_and_gates(
+        observable.ops, diagonalization_wires, unsupported_obs
+    )
+
+    new_observable = copy(observable)
+    new_observable._ops = new_ops
+
+    return diagonalizing_gates, new_observable
+
+
+@_diagonalize_complex_observable.register
+def _diagonalize_composite_op(
+    observable: CompositeOp, diagonalization_wires, unsupported_obs=_default_unsupported_obs
+):
+    diagonalizing_gates, new_operands = get_obs_and_gates(
+        observable.operands, diagonalization_wires, unsupported_obs
+    )
+
+    new_observable = copy(observable)
+    new_observable.operands = new_operands
```

Comment:
I wonder if swapping out `operands` of an observable will cause any issues, since it doesn't update things like `pauli_rep` or `grouping_indices`.

Reply:
Good point, I changed it to generate a new observable, so that _should_ be better